internal/vault/credentials/aws.go

```diff
@@ -0,0 +1,220 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package credentials
+
+import (
+	"context"
+	"strconv"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/uuid"
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/hashicorp/go-secure-stdlib/awsutil"
+	secretsv1alpha1 "github.com/hashicorp/vault-secrets-operator/api/v1alpha1"
+	"github.com/hashicorp/vault-secrets-operator/internal/consts"
+)
+
+const (
+	AWSAnnotationRole            = "eks.amazonaws.com/role-arn"
+	AWSAnnotationAudience        = "eks.amazonaws.com/audience"
+	AWSAnnotationTokenExpiration = "eks.amazonaws.com/token-expiration"
+	AWSDefaultAudience           = "sts.amazonaws.com"
+	AWSDefaultTokenExpiration    = int64(86400)
+)
+
+var _ CredentialProvider = (*AWSCredentialProvider)(nil)
+
+type AWSCredentialProvider struct {
+	authObj           *secretsv1alpha1.VaultAuth
+	providerNamespace string
+	uid               types.UID
+}
+
+func (l *AWSCredentialProvider) GetNamespace() string {
+	return l.providerNamespace
+}
+
+func (l *AWSCredentialProvider) GetUID() types.UID {
+	return l.uid
+}
+
+func (l *AWSCredentialProvider) Init(ctx context.Context, client ctrlclient.Client, authObj *secretsv1alpha1.VaultAuth, providerNamespace string) error {
+	l.authObj = authObj
+	l.providerNamespace = providerNamespace
+
+	if l.authObj.Spec.AWS.AWSCredsRef != "" {
+		// If AWSCredsRef is not empty, get the secret and read the creds from
+		// there, use the secret UID as l.uid
+		key := ctrlclient.ObjectKey{
+			Namespace: l.providerNamespace,
+			Name:      l.authObj.Spec.AWS.AWSCredsRef,
+		}
+		credsSecret, err := getSecret(ctx, client, key)
+		if err != nil {
+			return err
+		}
+		l.uid = credsSecret.UID
+	} else if l.authObj.Spec.AWS.IRSAServiceAccount != "" {
+		// Otherwise if the IRSA ref is not empty, read the service account, and
+		// use the service account UID as l.uid
+		key := ctrlclient.ObjectKey{
+			Namespace: l.providerNamespace,
+			Name:      l.authObj.Spec.AWS.IRSAServiceAccount,
+		}
+		irsaServiceAccount, err := getSA(ctx, client, key)
+		if err != nil {
+			return err
+		}
+		l.uid = irsaServiceAccount.UID
+	} else {
+		// At this point either the node role or the instance profile will be
+		// used for credentials, so just generate a new UID
+		l.uid = uuid.NewUUID()
```

Comment:
The only issue I see with this is that the UUID change when VSO is restarted, that will make it impossible for the client to ever restored from the storage cache. See https://github.com/hashicorp/vault-secrets-operator/pull/155#discussion_r1183097733 for a bit more  context.

Do either the node role or instance profile provide any unique ID we could use here?

Reply:
Hmm, I don't think so, since we wouldn't know which of the node role or instance profile to use here since that's configured on the Vault role side. Plus those would be the same for any VaultAuth in the cluster. 

Would a UUID constructed from a hash of the VaultAuth spec plus the k8s name and namespace suffice?

Reply:
Gave that a shot in 348a5aa

Reply:
Hmm, I don't think that will work in the case of the "default" VaultAuth. I think we would end up with the same hash for all tenants. I wonder if we really need to support this credential source, since it seems quite global? Unless I am missing something here?

Reply:
For the support question, it's just with the aim of supporting as many of the aws auth options as we can.

Reply:
Since this case is cluster-wide, would setting a static UID work here? i.e. `l.uid = types.UID("01234567-89ab-cdef-01234567")`

Otherwise I guess we just drop this case for now? Though it seems like it may be useful in the default VaultAuth case (i.e. the default operator auth uses the cluster's node role, then users configure their own for IRSA, etc.)

Reply:
Or instead of hard-coding a UID, we could use the `kube-root-ca.crt` configmap UID from the operator's namespace as something "cluster-wide". Or there's an `aws-auth` configmap in the `kube-system` namespace that is specific to EKS clusters, though I'm a little hesitant to read something from kube-system.